[CELEBORN-894] End to End Integrity Checks

gauravkm opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  - Make sure the PR title start w/ a JIRA ticket, e.g. '[CELEBORN-XXXX] Your PR title ...'.
  - Be sure to keep the PR description updated to reflect all changes.
  - Please write your PR title to summarize what this PR proposes.
  - If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
Design doc - https://docs.google.com/document/d/1YqK0kua-5rMufJw57kEIrHHGbLnAF9iXM5GdDweMzzg/edit?tab=t.0#heading=h.n5ldma432qnd

- End to End integrity checks provide additional confidence that Celeborn is producing complete as well as correct data
- The checks are hidden behind a client side config that is false by default. Provides users optionality to enable these if required on a per app basis
- Only compatible with Spark at the moment
- No support for Flink (can be considered in future)
- No support for Columnar Shuffle (can be considered in future)

Writer
- Whenever a mapper completes, it reports crc32 and bytes written on a per partition basis to the driver

Driver
- Driver aggregates the mapper reports - and computes aggregated CRC32 and bytes written on per partitionID basis

Reader
- Each CelebornInputStream will report (int shuffleId, int partitionId, int startMapIndex, int endMapIndex, int crc32, long bytes) to driver when it finished reading all data on the stream
- On every report
  - Driver will aggregate the CRC32 and bytesRead for the partitionID 
  - Driver will aggregate mapRange to determine when all sub-paritions have been read for partitionID have been read
  - It will then compare the aggregated CRC32 and bytes read with the expected CRC32 and bytes written for the partition
  - There is special handling for skewhandlingwithoutMapRangeSplit scenario as well
  - In this case, we report the number of sub-partitions and index of the sub-partition instead of startMapIndex and endMapIndex

There is separate handling for skew handling with and without map range split

As a follow up, I will do another PR that will harden up the checks and perform additional checks to add book keeping that every CelebornInputStream makes the required checks

### Why are the changes needed?
https://issues.apache.org/jira/browse/CELEBORN-894

Note: I am putting up this PR even though some tests are failing, since I want to get some early feedback on the code changes. 

### Does this PR introduce _any_ user-facing change?
Not sure how to answer this. A new client side config is available to enable the checks if required

### How was this patch tested?
Unit tests + Integration tests

FMX commented:
@gauravkm Hi, can you fix the CIs first? Lots of CI failures mean this PR has broken something.

gauravkm commented:
@FMX Thanks for looking! 

1. Yes. I am working on fixing the failures in parallel.
2. I put this up because i wanted some early feedback on the changes in `client/src/main/java/org/apache/celeborn/client/ShuffleClient.java`. I am adding an additional param to mapperEnd. This change seems to break some of the Flink tests (/home/runner/work/celeborn/celeborn/client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleOutputGate.java)

Specifically a couple of questions:
1. Is this too invasive of a change? 
2. Would you recommend keeping this change and fixing everything or undoing this and finding an alternate way to make this work?

gauravkm commented:
That makes sense. I will focus on getting a green build then.

buska88 commented on client/src/main/scala/org/apache/celeborn/client/commit/PartitionCompletenessValidator.scala:
```diff
@@ -0,0 +1,237 @@
+      skewPartitionHandlingWithoutMapRange: Boolean): (Boolean, String) = {
+    // TODO - ensure sub partitions aren't processed twice
+    if (skewPartitionHandlingWithoutMapRange) {
+      totalSubPartitionsProcessed.put(partitionId, totalSubPartitionsProcessed.get(partitionId) + 1)
```
will speculative task have influence on it?

gauravkm replied:
Good catch. 
I will de-dup based on the index

buska88 commented on client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:
```diff
@@ -1011,6 +1043,12 @@ public int pushOrMergeData(
 
+    // Track commit metadata if shuffle compression and integrity check are enabled and this request
+    // is not for pushing metadata itself.
+    if (shuffleCompressionEnabled && shuffleIntegrityCheckEnabled) {
```
I remember that in the first desigin(send commitMeta to worker), shuffleIntegrityCheckEnabled accuires shuffleCompressionEnabled enable;If we choose to send commitMeta to driver, should shuffleIntegrityCheckEnabled still need to be accuired?

gauravkm replied:
Is your question whether `shuffleIntegrityCheck` can be enabled without enabling compression?

1. `shuffleIntegrityCheckEnabled` is for enabling/disabling integrity checks on a per app basis
2. Shuffle Compression being enabled is not strictly required

gauravkm replied:
Removed

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -487,6 +521,31 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
+    if (isValid) {
+      response =
+        PbReducerPartitionEndResponse.newBuilder().setStatus(StatusCode.SUCCESS.getValue).build();
+    } else {
```
Ack. Will remove

gauravkm replied:
Removed

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/PartitionCompletenessValidator.scala:
```diff
@@ -0,0 +1,253 @@
+          bitmap = new RoaringBitmap();
+          totalSubPartitionsProcessed.put(partitionId, bitmap)
+        }
+        bitmap.add(endMapIndex)
```
Both good points. I can refactor to unify the branches.

gauravkm replied:
Updated to unify the branches

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala:
```diff
@@ -268,16 +276,40 @@ class ReducePartitionCommitHandler(
+            "commitMetadataArray can only be null if shuffleId %s is not registered",
+            shuffleId)
+          for (i <- 0 until numPartitions) {
+            if (crc32PerPartition(i) != 0) {
```
Oh interesting. Good point. I will update

gauravkm replied:
Updated

zaynt4606 commented on client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:
```diff
@@ -201,8 +208,10 @@ private static final class CelebornInputStreamImpl extends CelebornInputStream {
     private final boolean rangeReadFilter;
     private final boolean enabledReadLocalShuffle;
     private final String localHostAddress;
+    private final Map<String, CommitMetadata> expectedCommitMetadataMap = new HashMap<>();
```
Not in use.

gauravkm replied:
Removed

zaynt4606 commented on client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:
```diff
@@ -715,6 +735,50 @@ public synchronized void close() {
     }
 
+    void validateIntegrity() {
+      if (integrityChecked) {
```
```suggestion
      if (integrityChecked || !shuffleIntegrityCheckEnabled ) {
        return;
```
There could be a lot of logs here because there could be a lot of InputStreams

gauravkm commented on common/src/main/java/org/apache/celeborn/common/write/PushState.java:
```diff
@@ -33,6 +34,8 @@ public class PushState {
   private final int pushBufferMaxSize;
   public AtomicReference<IOException> exception = new AtomicReference<>();
   private final InFlightRequestTracker inFlightRequestTracker;
+  private final ConcurrentHashMap<Integer, CommitMetadata> commitMetadataMap =
```
Added

gauravkm commented:
@FMX @buska88 @gaoyajun02 @zaynt4606 Thanks for the feedback until now.

I have fixed all the tests and addressed all existing feedback. Please do take a look when you get a chance. Thanks!

FMX commented:
> @FMX @buska88 @gaoyajun02 @zaynt4606 Thanks for the feedback until now.
> 
> I have fixed all the tests and addressed all existing feedback. Please do take a look when you get a chance. Thanks!

Thank you for fixing all the UTs. The review will be done within this week.

gauravkm commented on client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:
```diff
@@ -647,6 +650,35 @@ public Tuple2<Integer, Boolean> getShuffleId(
+            pbReducerPartitionEnd,
+            conf.clientRpcRegisterShuffleAskTimeout(),
+            ClassTag$.MODULE$.apply(PbReducerPartitionEndResponse.class));
+    if (pbReducerPartitionEndResponse.getStatus() != StatusCode.SUCCESS.getValue()) {
```
Looking at the code it seems like in CelebornInputStream, there is existing handling for throwing `FetchFailedException` in case of exceptions during `fillBuffer`.

So if I throw CelebornIOException directly (instead of wrapping in `RuntimeException`) during `validateIntegrity`, it will end up hitting the existing codepath.

gauravkm replied:
I have updated the code in `validateIntegrity` to not wrap the exception to `RuntimeException`

gauravkm replied:
Actually, the comment in code says
```
/*
             * [[ExceptionMaker.makeException]], for spark applications with celeborn.client.spark.fetch.throwsFetchFailure enabled will result in creating
             * a FetchFailedException; and that will make the TaskContext as failed with shuffle fetch issues - see SPARK-19276 for more.
             * Given this, Celeborn can wrap the FetchFailedException with our CelebornIOException
             */
```

Should i just plug into this or explicitly rely on `stage.rerun.enabled` config? I don't see many uses of it apart from LifecycleManager

gauravkm replied:
`CelebornIntegrityCheckSuite` exercises these code paths and it works as expected. Resolving

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidator.scala:
```diff
@@ -0,0 +1,75 @@
+      if (bitmap.contains(endMapIndex)) {
+        // check if previous entry matches
+        if (subPartitionToCommiMetadata.get(endMapIndex) != actualCommitMetadata) {
+          (false, "")
```
good catch. Fixed

gauravkm commented on common/src/main/java/org/apache/celeborn/common/write/PushState.java:
```diff
@@ -102,4 +106,35 @@ public void recordFailedBatch(String partitionId, int mapId, int attemptId, int
+
+  public long[] getBytesWrittenPerPartition(
+      boolean shuffleIntegrityCheckEnabled, int numPartitions) {
+    long[] bytesWrittenPerPartition = new long[numPartitions];
```
Updated to empty array

gauravkm commented on common/src/main/java/org/apache/celeborn/common/write/PushState.java:
```diff
@@ -102,4 +106,35 @@ public void recordFailedBatch(String partitionId, int mapId, int attemptId, int
+
+  public int[] getCRC32PerPartition(boolean shuffleIntegrityCheckEnabled, int numPartitions) {
+    int[] crc32PerPartition = new int[numPartitions];
+    if (!shuffleIntegrityCheckEnabled) {
```
Updated to empty array

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:
```diff
@@ -438,6 +456,22 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
           throw new UnsupportedOperationException(s"Not support $partitionType yet")
       }
 
+    case pb: ReducerPartitionEnd =>
```
Ack. Updating

gauravkm commented on client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:
```diff
@@ -715,6 +734,41 @@ public synchronized void close() {
+            key,
+            aggregatedActualCommitMetadata);
+      } else {
+        shuffleClient.reducerPartitionEnd(
```
ack. Updating to `readReducerPartitionEnd` as suggested

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/LegacySkewHandlingPartitionValidator.scala:
```diff
@@ -0,0 +1,173 @@
+      if (sumOfMapRanges > expectedTotalMapperCount) {
+        val errorMsg = s"AQE Partition $partitionId failed validation check " +
+          s"while processing startMapIndex: $startMapIndex endMapIndex: $endMapIndex " +
+          s"ActualCommitMatadata $currentCommitMetadata > ExpectedTotalMapperCount $expectedTotalMapperCount"
```
thanks. Fixing

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/LegacySkewHandlingPartitionValidator.scala:
```diff
@@ -0,0 +1,173 @@
+  private val partitionToSubPartitionCount = JavaUtils.newConcurrentHashMap[Int, Int]()
+  private val currentCommitMetadataForReducer =
+    JavaUtils.newConcurrentHashMap[Int, CommitMetadata]()
+  private val currentTotalMapRangeSumForReducer = JavaUtils.newConcurrentHashMap[Int, Int]()
```
Ack. Renaming

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidator.scala:
```diff
@@ -0,0 +1,79 @@
+
+  private val totalSubPartitionsProcessed = JavaUtils.newConcurrentHashMap[Int, RoaringBitmap]()
+  private val partitionToSubPartitionCount = JavaUtils.newConcurrentHashMap[Int, Int]()
+  private val subPartitionToCommiMetadata = JavaUtils.newConcurrentHashMap[Int, CommitMetadata]()
```
Fixing

gauravkm commented:
> I think we can support the Flink engines later, maybe in the following PRs.

Yeah. Definitely,

FMX commented:
> @FMX @RexXiong Thanks for the review. I have updated the PR with review changes

Have you tested this PR on a Celeborn cluster to check if there are any changes in performance and stability? Especially when there are many partition locations, the driver's memory can be stable.

gauravkm commented:
>Have you tested this PR on a Celeborn cluster to check if there are any changes in performance and stability? Especially when there are many partition locations, the driver's memory can be stable.

Good question!
* Yes, we have been running 1000's of applications within Stripe with a similar implementation for the last 3 months
* Our default partition count is 5k, and we have apps with 100k-600k mappers in a single stage (and multiple such stages) that have been running reliably and performantly 
* Internally at Stripe - we enable these checks by default, so every app runs with these checks
* The additional memory footprint on the driver is independent of number of PartitionLocations and only depends on the number of partitions * number of stages in the app 

Per stage additional memory can be computed as - number of partitions * 12 bytes (4 for crc32, 8 for byte count) 
So for 5k partitions -> 60KB

- Assuming even 100 such stages, the additional memory footprint for the driver is 6MB. We haven't observed any driver OOM issues etc with this implementation
- Apart from this if the app has skewed partitions, there is some additional tracking of sub-partition information. Even assuming 1000's of sub-partitions and 100 additional bytes per sub-partition, we are looking at 100 KB of additional memory

* With respect to perf, our benchmarks show that for large apps shuffle time dominates the task time anyway and the overhead from these checks is minimal. We didn't measure any meaningful increase in overall Spark app run time due to these checks.
  - If you look at the [design](https://docs.google.com/document/d/1YqK0kua-5rMufJw57kEIrHHGbLnAF9iXM5GdDweMzzg/edit?tab=t.0#heading=h.n5ldma432qnd), the first iteration was very different. We found perf issues with that implementation in certain scenarios (specifically when the app has a lot of empty partitions) and the v2/v3 design eliminated those performance concerns

gauravkm commented:
<img width="1364" alt="Screenshot 2025-06-05 at 7 29 09 PM" src="https://github.com/user-attachments/assets/bc5b8d59-3eef-453b-9cc1-a89e4103d41f" />

gauravkm commented:
Additionally, the feature is behind a feature flag, so it does not change the status quo.

For users who try this out, If we get user reports of bottlenecks/hotspots/instability while using this feature, then I am happy to debug and improve this feature

RexXiong commented on client/src/main/java/org/apache/celeborn/client/read/CelebornInputStream.java:
```diff
@@ -721,6 +740,36 @@ public synchronized void close() {
+      }
+
+      if (readSkewPartitionWithoutMapRange) {
+        shuffleClient.readReducerPartitionEnd(
```
The current implementation mixes the concepts of numberOfSubPartitions/currentIndexOfSubPartition and startMapIndex/endMapIndex across different method signatures, such as readReducerPartitionEnd(int startMapIndex, int endMapIndex) and readPartition(int startMapIndex, int endMapIndex). Meanwhile, CelebornInputStreamImpl utilizes both. To ensure consistency, IMO we should use (int startMapIndex, int endMapIndex), as seen in CelebornShuffleReader.

gauravkm replied:
The existing Celeborn implementation in `CelebornInputStream` explicitly sets the `startMapIndex` and `endMapIndex` to -1 in the case of `skewPartitionHandlingWithoutMapRange`. That is why i left the existing convention alone.
Would you know why the original implementation chose to set the values to -1?

RexXiong replied:
> I propose we keep this as is for now. If required, we can always make a smaller change in a follow up PR. Will make it easier to reason about as well

Agree, this can be done in next prs

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:
```diff
@@ -5356,6 +5358,14 @@ object CelebornConf extends Logging {
       .bytesConf(ByteUnit.BYTE)
       .createWithDefaultString("512k")
 
+  val CLIENT_SPARK_SHUFFLE_INTEGRITY_CHECK_ENABLED: ConfigEntry[Boolean] =
```
Maybe better to revert back to using `celeborn.client.shuffle.integrityCheck.enabled` , as the configuration use in many place not only impact spark, As a compromise, just comment this configuration is only affects spark

gauravkm replied:
Ack. Updating

RexXiong commented on client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala:
```diff
@@ -82,6 +83,13 @@ class ReducePartitionCommitHandler(
 
+  private val shuffleIntegrityCheckEnabled = conf.clientShuffleIntegrityCheckEnabled
+  // partitionId-shuffleId -> number of mappers that have written to this reducer (partition + shuffle)
+  private val commitMetadataForReducer =
```
Should clean `commitMetadataForReducer/aqePartitionCompletenessValidator` when the shuffle expired

gauravkm replied:
Ack. Removing in `removeExpiredShuffle()`

gauravkm commented on common/src/main/java/org/apache/celeborn/common/CommitMetadata.java:
```diff
@@ -0,0 +1,88 @@
+    return new CommitMetadata(checksum, numBytes);
+  }
+
+  public static boolean checkCommitMetadata(CommitMetadata expected, CommitMetadata actual) {
```
This one does get used for comparing the metadata in ReducePartitionCommitHandler and PartitionCompletenessValidator

gauravkm commented on client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:
```diff
@@ -203,13 +203,19 @@ public abstract int mergeData(
 
-  // Report partition locations written by the completed map task of ReducePartition Shuffle Type
-  public abstract void mapperEnd(int shuffleId, int mapId, int attemptId, int numMappers)
+  // Report partition locations written by the completed map task of MapPartition Shuffle Type.
```
Not on purpose. Reverting

gauravkm commented:
@RexXiong @FMX Thanks for the review! I have updated the PR with the recommended changes.

RexXiong reviewed (approved):
LGTM, Could we add some negative case for this feature like `CelebornFetchFailureSuite`?

gauravkm commented:
> LGTM, Could we add some negative case for this feature like CelebornFetchFailureSuite?

Thanks for the review!
Yes, I can add.
Do you recommend adding it as part of this PR or a follow up PR?

gauravkm commented:
> What will happen if split: HARD_SPLIT happened?
These checks work end to end. It doesn't matter which workers get written to. As long the reducers read the requisite number of bytes and crc32 matches, data will be accurate

> What mechanism guarantees the integrity of data sent from the client to the worker?
Celeborn already allows you to enable shuffle data compression that guarantees the integrity of every package exchanged between the client and worker

gauravkm commented:
@RexXiong Please take a look at `CelebornIntegrityCheckSuite`. 

1. I introduced a bit flip in one of the partition location files using `ShuffleReaderGetHookForCorruptedData`
2. Then test for 3 scenarios:
- Run app without integrity checks - App reads expected number of records, app succeeds but data does not match
- Run app with integrity checks and no stage rerun - As expected app fails due to commit metadata mismatch
- Run app with integrity checks and stage rerun enabled - App succeeds and data matches

gaoyajun02 commented on client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala:
```diff
@@ -301,19 +334,71 @@ class ReducePartitionCommitHandler(
         0 until numMappers foreach (idx => attempts(idx) = -1)
         shuffleMapperAttempts.put(shuffleId, attempts)
       }
+      commitMetadataForReducer.put(shuffleId, new Array[CommitMetadata](numPartitions))
```
Should we add a shuffleIntegrityCheckEnabled check here?

gauravkm replied:
Added

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/LegacySkewHandlingPartitionValidator.scala:
```diff
@@ -0,0 +1,180 @@
+        override def apply(key: Int): util.TreeMap[(Int, Int), CommitMetadata] =
+          new util.TreeMap[(Int, Int), CommitMetadata](comparator)
+      })
+    subRangeToCommitMetadataMap.synchronized {
```
Fixed

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/LegacySkewHandlingPartitionValidator.scala:
```diff
@@ -0,0 +1,180 @@
+          currentCommitMetadataForReducer.merge(
+            partitionId,
+            actualCommitMetadata,
+            new java.util.function.BiFunction[CommitMetadata, CommitMetadata, CommitMetadata] {
```
Fixed

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/PartitionCompletenessValidator.scala:
```diff
@@ -0,0 +1,87 @@
+      actualCommitMetadata: CommitMetadata,
+      expectedCommitMetadata: CommitMetadata,
+      expectedTotalMapperCountForParent: Int,
+      skewPartitionHandlingWithoutMapRange: Boolean): (Boolean, String) = {
```
Thanks. Updated

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/LegacySkewHandlingPartitionValidator.scala:
```diff
@@ -0,0 +1,180 @@
+      startMapIndex: Int,
+      endMapIndex: Int,
+      actualCommitMetadata: CommitMetadata,
+      expectedTotalMapperCount: Int): (Boolean, String) = {
```
Thanks. Updated

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidator.scala:
```diff
@@ -0,0 +1,79 @@
+        bitmap.add(endMapIndex)
+        subPartitionToCommitMetadata.put(endMapIndex, actualCommitMetadata)
+      }
+    }
```
Yeah we can consider additional argument validation to ensure that the invariant holds

gauravkm replied:
Added argument validation at beginning of the method

buska88 replied:
Oh, let me explain it. If startMapIndex equals 10, but SkewHandlingWithoutMapRangeValidator receives 11 different endMapIndex, what should we do?Or do u think such cases should be handle in this pr?

gauravkm replied:
Updated. Added additional state check

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/CommitHandler.scala:
```diff
@@ -427,7 +427,9 @@ abstract class CommitHandler(
+  /**
+   * Invoked when a reduce partition finishes reading data to perform end to end integrity check validation
+   */
    */
```
No worries. updated

gauravkm commented:
@gaoyajun02 I have addressed all the feedback. Please do take another look when you get a chance.

(I will fix any test failures in case anything broke with the recent changes)

gaoyajun02 reviewed (approved):
LGTM, thanks. I can also take a look at the test code if I have time tonight.

gaoyajun02 commented:
> * we have apps with **100k-600k mappers in a single stage (and multiple such stages)** that have been running reliably and performantly
Additionally, we still have concerns about high concurrency scenarios (executor * cores). We have actually applied end-to-end consistency validation in production scenarios, and recently we've been analyzing cases of driver celeborn RPC timeouts (500k mappers, 8000*2 cores), which may not necessarily be related to this change.

buska88 commented:
> > * we have apps with **100k-600k mappers in a single stage (and multiple such stages)** that have been running reliably and performantly
> 
> Additionally, we still have concerns about high concurrency scenarios (executor * cores). We have actually applied end-to-end consistency validation in production scenarios, and recently we've been analyzing cases of driver celeborn RPC timeouts (500k mappers, 8000*2 cores), which may not necessarily be related to this change.

We test this job in several cases.
If we set spark.celeborn.client.shuffle.integrityCheck.enabled=false,  a MapEnd request costs 0.1-0.2ms;
Use current pr to test, a MapEnd request costs 50-70ms.
We find that the bottleneck is finishMapperAttempt in ReducePartitionCommitHandler, for (i <- 0 until numPartitions) {} costs a lot when numPartitions and numMappers is big, and if we choose to put it in shuffleMapperAttempts.synchronized, it leads to long MapEnd request.
@gauravkm @gaoyajun02

gauravkm commented on client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala:
```diff
@@ -268,16 +277,40 @@ class ReducePartitionCommitHandler(
+            }
+          }
+        }
+
```
Thanks. That makes sense. Updated

buska88 commented:
> Thanks for testing!
> 
> > when numPartitions and numMappers is big
> 
> What do the concrete values for these look like in your test?

<img width="737" alt="image" src="https://github.com/user-attachments/assets/be8ccb28-dac5-4d50-aab3-18a0196db6a2" />
the job has 2 big stages with 400000 and 800000 map tasks, and it's numpartitions is 6000

gaoyajun02 commented on client/src/main/scala/org/apache/celeborn/client/commit/ReducePartitionCommitHandler.scala:
```diff
@@ -296,24 +309,94 @@ class ReducePartitionCommitHandler(
+          val commitMetadata = commitMetadataArray(i)
+          commitMetadata.addCommitData(
+            crc32PerPartition(i),
+            bytesWrittenPerPartition(i))
```
commitMetadataArray(i).addCommitData(crc32PerPartition(i), bytesWrittenPerPartition(i))

gaoyajun02 commented on client/src/test/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidatorTest.scala:
```diff
@@ -0,0 +1,257 @@
+    try {
+      validator.processSubPartition(partitionId, startMapIndex, endMapIndex, metadata, 20)
+    } catch {
+      case e: IllegalArgumentException => e.isInstanceOf[IllegalArgumentException] shouldBe true
```
Consider using `intercept[IllegalStateException]` instead of try-catch for more concise and precise exception testing.

gaoyajun02 commented on client/src/test/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidatorTest.scala:
```diff
@@ -0,0 +1,257 @@
+    try {
+      validator.processSubPartition(partitionId, 12, 1, metadata, 12)
+    } catch {
+      case e: IllegalStateException => e.isInstanceOf[IllegalStateException] shouldBe true
```
ditto

gaoyajun02 commented on client/src/test/scala/org/apache/celeborn/client/commit/SkewHandlingWithoutMapRangeValidatorTest.scala:
```diff
@@ -0,0 +1,257 @@
+    val subPartitions3 = 3
+
+    // Process all sub-partitions for partition1
+    val expectedMetadata1 = new CommitMetadata();
```
```suggestion
    val expectedMetadata1 = new CommitMetadata()
```

gaoyajun02 reviewed (approved):
LGTM

gauravkm commented on client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:
```diff
@@ -648,6 +648,35 @@ public Tuple2<Integer, Boolean> getShuffleId(
   }
 
+  @Override
+  public void readReducerPartitionEnd(
```
Sure. Added

gauravkm replied:
If this PR looks good to you, then could you please approve it with "LGTM" message as well.

buska88 replied:
LGTM

RexXiong reviewed (commented):
LGTM, only a minor

gauravkm commented:
Thanks a lot @RexXiong @FMX @gaoyajun02 @buska88 @zaynt4606 for your support in getting this feature merged into Celeborn!

byteroll commented:
hi @gauravkm , I reproduced this bug https://github.com/apache/celeborn/pull/3507 and found that e2e check cannot detect the data loss, because the the check point is behind when the loss happend. Is this scenario within the scope of the E2E Check?

gauravkm commented:
@byteroll Thanks for reporting and reproducing the bug. I did not understand what you meant by “checkpoint is behind”
The scenario of shuffle client dropping data silently is indeed within e2e integrity check scope.

couple of questions:
1. Could you please elaborate a bit more on why the check did not happen?
2. It seems like the mapper OOM’ed but was still able to report success to the driver?

byteroll commented:
@gauravkm  I drew a schematic to illustrate it. 
https://mermaid.ai/d/71d0fc9c-0c6d-4696-b694-17f149d0823b

gauravkm commented:
@byteroll Thanks for the diagram

It still doesn't answer the questions I had and i have a few more:
1. Which version of end to end checksums are you using? In the latest version, checksum information is propagated through the driver. 
2. If you are using the version where metadata is sent as data packet, then why does the metadata package get delivered to worker if the Data Pusher thread stopped
3. Is the mapper marked successful or failed? If it is marked failed, then why does the reader try to read its data
4. Even if none of the above holds, your diagram shows that the mapper should have the correct expected checksum, since the checksum is computed on raw data. If the reader does not read the full data (because the mapper never pushed it fully), then i would expect both byte count check and crc checks to fail

Edit 1: I can understand why 4 succeeded from your diagram. So maybe you can answer the other 3

Edit 2: I understand the core issue. E2e checks assumed that data is making to ShuffleClient. But here the data is being dropped prior to hitting the ShuffleClient

byteroll commented:
Exactly. The "Edit 2" is what I mean.
And the pushThread is a daemon, so before fixed, the main thread will not be aware of the exception

gauravkm commented:
Can you tell me if the mapper was marked succeeded in this scenario?

gauravkm commented:
A potential fix here is to move the metadata computation to the HashBasedShuffleWriter and SortBasedShuffleWriter

byteroll commented:
> Can you tell me if the mapper was marked succeeded in this scenario?

I tried again just now, it shows success.

byteroll commented:
> A potential fix here is to move the metadata computation to the HashBasedShuffleWriter and SortBasedShuffleWriter

yes, the the metadata computation should be moved to the ShuffleWriter#write phrase, so it can cover all data.

gauravkm commented:
Makes sense. I will evaluate and put up a fix to move it to the Writers

gauravkm commented:
> I tried again just now, it shows success

Is the scenario such that only one thread is OOMing?

Are there specific steps to reproduce this issue?

byteroll commented:
Actually, as shown in diagram above, an unknown error caused by a bug can occur at any stage from the start of data writing  to the computation of the CRC32. I think the CRC32 should be computed from the start.

Here are my reproduce steps.


```
CREATE TABLE <XXX> AS
SELECT <field>
FROM <YYY>
DISTRIBUTE BY <field>;
```

celeborn-2176.patch
```
diff --git a/client/src/main/java/org/apache/celeborn/client/write/DataPusher.java b/client/src/main/java/org/apache/celeborn/client/write/DataPusher.java
index 73af67b07..a74f5980a 100644
--- a/client/src/main/java/org/apache/celeborn/client/write/DataPusher.java
+++ b/client/src/main/java/org/apache/celeborn/client/write/DataPusher.java
@@ -22,6 +22,7 @@ import java.util.ArrayList;
 import java.util.Objects;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.atomic.LongAdder;
 import java.util.concurrent.locks.Condition;
@@ -41,6 +42,39 @@ import org.apache.celeborn.common.write.PushState;
 public class DataPusher {
   private static final Logger logger = LoggerFactory.getLogger(DataPusher.class);

+  /**
+   * CELEBORN-2176 bug reproduction injection.
+   *
+   * <p>Set the system property {@code celeborn.test.dataPusher.injectErrorOnPush=N} to throw a
+   * RuntimeException on the N-th pushData call (1-based). This simulates a non-IOException escaping
+   * the push thread without being caught, silently stopping the thread and causing data loss.
+   *
+   * <p>Optionally restrict injection to a specific mapId range:
+   *
+   * <ul>
+   *   <li>{@code celeborn.test.dataPusher.injectErrorMapIdStart} — first mapId to inject (inclusive)
+   *   <li>{@code celeborn.test.dataPusher.injectErrorMapIdEnd} — last mapId to inject (inclusive)
+   * </ul>
+   *
+   * <p>Example: inject on the 1st push of mapId 2..4:
+   *
+   * <pre>
+   *   -Dceleborn.test.dataPusher.injectErrorOnPush=1
+   *   -Dceleborn.test.dataPusher.injectErrorMapIdStart=2
+   *   -Dceleborn.test.dataPusher.injectErrorMapIdEnd=4
+   * </pre>
+   */
+  private static final int INJECT_ERROR_ON_PUSH =
+      Integer.getInteger("celeborn.test.dataPusher.injectErrorOnPush", -1);
+
+  private static final int INJECT_ERROR_MAP_ID_START =
+      Integer.getInteger("celeborn.test.dataPusher.injectErrorMapIdStart", -1);
+
+  private static final int INJECT_ERROR_MAP_ID_END =
+      Integer.getInteger("celeborn.test.dataPusher.injectErrorMapIdEnd", -1);
+
+  private final AtomicInteger pushCallCount = new AtomicInteger(0);
+
   private final long WAIT_TIME_NANOS = TimeUnit.MILLISECONDS.toNanos(500);

   private LinkedBlockingQueue<PushTask> idleQueue;
@@ -139,10 +173,10 @@ public class DataPusher {
                   } catch (InterruptedException e) {
                     logger.error("DataPusher push thread interrupted while pushing data.");
                     break;
-                  } catch (Throwable e) {
-                    logger.error("Unexpected exception occurs.", e);
-                    exceptionRef.set(new IOException(e));
                   }
+                  // CELEBORN-2176 bug reproduction: catch (Throwable) is intentionally removed.
+                  // Non-IOException thrown in pushData will silently kill this thread without
+                  // setting exceptionRef, causing data loss.
                 }
               }
             },
@@ -208,6 +242,19 @@ public class DataPusher {
   }

   protected void pushData(PushTask task) throws IOException {
+    boolean mapIdMatches =
+        INJECT_ERROR_MAP_ID_START < 0
+            || (mapId >= INJECT_ERROR_MAP_ID_START && mapId <= INJECT_ERROR_MAP_ID_END);
+    if (INJECT_ERROR_ON_PUSH > 0
+        && mapIdMatches
+        && pushCallCount.incrementAndGet() == INJECT_ERROR_ON_PUSH) {
+      throw new RuntimeException(
+          "[CELEBORN-2176 reproduction] Injected non-IOException in pushData call #"
+              + INJECT_ERROR_ON_PUSH
+              + " (mapId="
+              + mapId
+              + ") to simulate push thread dying silently");
+    }
     int bytesWritten =
         client.pushData(
             shuffleId,
```
```
"spark.executor.extraJavaOptions": "-Dceleborn.test.dataPusher.injectErrorOnPush=1 -Dceleborn.test.dataPusher.injectErrorMapIdStart=2 -Dceleborn.test.dataPusher.injectErrorMapIdEnd=4"
```

gauravkm commented:
1. Checks at the Writer layer are certainly better and Data Pusher erroring out without the task failing is a clear gap
2. I am just trying to understand the specific OOM scenario that you originally encountered since OOMs are common and more tangible than other unknown errors.


In the cases where mapper is marked successful, the logging change from your original PR is a no-op then?

byteroll commented:
Sorry, I’m not a native English speaker, and I didn’t understand this sentence:

> the logging change from your original PR is a no-op then?

I didn’t reproduce the OOM issue; I just injected a random exception in the pushThread to simulate a data loss scenario.
And it works, spark metrics shows that the shuffle writen records != records read, and the jobs completed successfully.

```
Exchange
shuffle records written: 356,984
shuffle write time total (min, med, max (stageId: taskId))
1.1 s (245 ms, 351 ms, 458 ms (stage 0.0: task 2))
records read: 331,697
fetch wait time: 899 ms
remote bytes read: 11.0 MiB
remote blocks read: 3,275
data size total (min, med, max (stageId: taskId))
76.8 MiB (20.6 MiB, 28.0 MiB, 28.2 MiB (stage 0.0: task 0))
number of partitions: 1,000
shuffle bytes written total (min, med, max (stageId: taskId))
11.0 MiB (2.7 MiB, 4.0 MiB, 4.2 MiB (stage 0.0: task 0))
```

gauravkm commented:
Thanks!

My understanding is that if the mapper is marked successful, then the logging change from your DataPusher fix PR that logs unsuccessful mapper attempts, won’t get triggered. Is that accurate?

thanks for reporting the issue. I really appreciate it.

byteroll commented:
Oh, you mean this line https://github.com/apache/celeborn/pull/3507/changes#diff-11c5cfca6536590e8f2c8ecadb08bfcb4ccad743328565c637defa10930e690eR351, the log info from the bug fix pr ?
I'll check it tomorrow.

byteroll commented:
> My understanding is that if the mapper is marked successful, then the logging change from your DataPusher fix PR that logs unsuccessful mapper attempts, won’t get triggered. Is that accurate?

yeah, it was not triggered.
